### ISSUE TITLE
[Feature idea] Individual video quality setting per podcast.

### ISSUE THREAD

**Issue by Marrib** [gpodder/gpodder#1164]
gPodder allows to set the quality of the videos to be downloaded globally. But it would be very practical and handy if video download quality could be set individually per channel.

By default when a channel is added it could/should use the global quality setting. But there should be an option in the channel setting dialog to override the global quality setting and to chose a channel-specific setting.

For YouTube channels added as podcasts this implementation would be easy, I guess, because different video quality download links are included in one video URL.

For podcast channels added via an XML (or similar) file which include only one quality, I propose this solution: There should be an option in the channel settings dialog to add multiple Feed URLs. gPodder can then download the video quality set by the user from the according Feed URL.

**Comment by auouymous** [comment]
I too want per-channel settings, and it has been on my list for a while now, but we first need to implement #1032, because format IDs only work for youtube and channels supported by youtube-dl. And manually adding format IDs is too complicated for most users.

Some non-youtube feeds contain multiple enclosures with different audio or video formats. I'm not sure how podcastparser or gpodder handles such feeds, it might discard all but one. Keeping those enclosures would allow a global or per-channel audio bitrate and video resolution to select one of them.

As for multiple feed URLs per channel, I know such podcasts exist, and the user must choose a format when subscribing. There should however be an option to change a feed's URL (#214) so the user could chose another format, in addition to updating a moved feed. But changing the feed URL when your connection changes is complicated, see that PR for some reasons. And once changed, gpodder would need to force update the feed which changes the enclosure for every episode. This wouldn't cause any issues with already downloaded episodes, but it is an additional complication.

**Comment by Marrib** [comment]
Thanks for that information. Good to know that such amazing new functionality is in the making... maybe slowly, but in the making.

One little misunderstanding, though, because I didn't express my thought as precisely and in a good descriptive way as I should have:
I didn't mean that gPodder should automatically change video/audio download quality based on internet speed or internet connection type. What I meant was the following: A user could set podcast A to download 1080p video, and podcast B to download 540p video. This way he could download different qualities at the same time. Without the need to manually change the global setting every time between 1080p and 540p.

**Comment by auouymous** [comment]
Automatically selecting quality based on network speed is a bad idea. Your gigabit connection might be temporarily congested during an update, and all episodes are downloaded at 360p instead of a higher resolution. The advantage of using gpodder is that it can slowly download a high resolution video in the background, unlike streaming which might constantly buffer.

What I meant was that allowing a user to manually switch feed URLs for a single podcast based on their current bandwidth has too many complications. The whole per-channel settings idea is completely doable, but not for podcasts that use separate feeds for each format.

But manually changing the global setting when on a metered connection might not work as expected, unless there is a global option to override per-channel settings. Imagine the global setting is 720p, but one channel uses 4k. Changing to 480p globally on a metered connection would still download that one channel at 4k. However, that override should only alter per-channel settings that are higher than the global. So a 360p channel will still download at 360p and not at the global 480p. And same goes for a youtube channel set to only download audio, it won't suddenly download the 480p video.

**Comment by Marrib** [comment]
> Automatically selecting quality based on network speed is a bad idea. Your gigabit connection might be temporarily congested during an update, and all episodes are downloaded at 360p instead of a higher resolution. The advantage of using gpodder is that it can slowly download a high resolution video in the background, unlike streaming which might constantly buffer.

Good point, I agree.
> What I meant was that allowing a user to manually switch feed URLs for a single podcast based on their current bandwidth has too many complications. The whole per-channel settings idea is completely doable, but not for podcasts that use separate feeds for each format.

How about a function like that: When a user adds multiple URLs for a single podcast, gPodder combines/re-parses both URLs to a single feed. This way users would have only one podcast in the podcast list on the left side, but two qualities in the episode list to choose for downloading for each episode. Which one of them is of higher quality can bee seen in the filesize column. The auto-download functionality/policy of gPodder, if activated, in this case would need to have a setting to download either the higher or lower quality file.
> But manually changing the global setting when on a metered connection might not work as expected, unless there is a global option to override per-channel settings. Imagine the global setting is 720p, but one channel uses 4k. Changing to 480p globally on a metered connection would still download that one channel at 4k. However, that override should only alter per-channel settings that are higher than the global. So a 360p channel will still download at 360p and not at the global 480p. And same goes for a youtube channel set to only download audio, it won't suddenly download the 480p video.

I agree, too.

**Comment by tpikonen** [comment]
@Marrib Do you have an example feed where there are multiple URLs, one for each quality setting? I'm sure they exist, I just have never seen one.

**Comment by Marrib** [comment]
Sure, here for example is the URL to a German public service broadcasting news institution:
https://www.tagesschau.de/infoservices/podcast

**Comment by tpikonen** [comment]
Thanks. It would be easier to implement support for a single feed with multiple files under a media:group tag, but in this case it's understandable that they want to have their feeds as simple as possible.

**Comment by Marrib** [comment]
Indeed, there are more clever methods out there, for example to create a single feed with multiple files under a media:group tag or similar concepts. And the mentioned news institution -- just like all the other podcast creators in question -- could probably do that as well. But they don't. So the solution to this problem probably must be on the software side of things.

**Comment by Marrib** [comment]
Maybe an approach like this would be reasonable, for example:

1) A user adds three feed-URLs (540p, 720p, 1080p) for the same podcast in the channel preference dialog.
2) gPodder then behind the scenes combines/reparses them to create a new feed with all the included episodes of all qualities (three entries per episode).
3) But instead of showing _all three_ entries per episode in the episodes list, only the _one_ entry according to the user-selected quality level is shown in the episodes list, say 720p.
4) The user then downloads one episode with 720p quality.
5) Then the user manually switches the quality setting of the podcast to 1080p. In this case the corresponding episode entries of this new quality are shown in the episodes list. And to avoid redundant downloading of the same episode in different quality levels, in this case the already downloaded 720p episode should also show up in the 1080p list.

**Comment by tpikonen** [comment]
An extension which multiplexes extra feeds together with a main feed into a single feed with a media:group tag would perhaps be possible (not sure if the extension mechanism has all the necessary hooks, but they could be added). The media:group support (#1032) needs to be implemented first.

Since there are not that many feeds like your example, I don't think this will be a high priority for anyone.

**Comment by auouymous** [comment]
I imagine the podcasts using multiple feeds are old and media:group probably didn't exist back then. Switching would not only cost money but might inconvenience users.
